client/state/comments/utils.js

```diff
@@ -0,0 +1,114 @@
+ * @returns {String} request id
+ */
+export function createRequestId( siteId, postId, query ) {
+	const queryKeys = Object.keys( query ).sort();
+	const queryString = queryKeys.map( ( key ) => `${ key }=${ query[key] }` ).join( '-' );
+
+	return `${siteId}-${postId}-${ queryString }`;
+}
+
+/***
+ * Builds a comment tree of the shape
+ * Map<id, CommentNode> {
+ * 	children: List<id>, // Array of root level comments ids
+ * }
+ * @param {Immutable.List} comments list of comments (as built on state.comments.items) sorted by date in descending order
+ * @returns {Immutable.Map} Immutable map comments tree instance of the shape Map<id, CommentNode>{ children: List<id> }
+ */
+export function buildCommentsTree( comments ) {
+	const aTree = Immutable.fromJS( { children: [] } );
+
+	return aTree.withMutations( ( commentsTree ) => {
+		comments.forEach( ( comment ) => {
+			// if the comment has a parent, but we haven't seen that parent yet, create a placeholder
+			if ( comment.get( 'parent' ) && ! commentsTree.has( comment.getIn( [ 'parent', 'ID' ] ) ) ) {
+				commentsTree.set( comment.getIn( [ 'parent', 'ID' ] ), Immutable.fromJS( {
+					children: [],
+					parentId: undefined,
+					data: undefined
+				} ) );
+			}
+
+			// if it's the first time we see that comment, create it
+			if ( ! commentsTree.has( comment.get( 'ID' ) ) ) {
+				commentsTree.set( comment.get( 'ID' ), Immutable.fromJS( {
+					children: [],
+					parentId: comment.get( 'parent' ) ? comment.getIn( [ 'parent', 'ID' ] ) : null,
+					data: comment
+				} ) );
+			} else {
+				// we already saw that comment, so it means it's parent of some other comment, so it already has children
+				// array, so we need only update the parentId and the data
+				const parentId = comment.get( 'parent' ) ? comment.getIn( [ 'parent', 'ID' ] ) : null;
+
+				commentsTree.setIn( [ comment.get( 'ID' ), 'parentId' ], parentId );
+				commentsTree.setIn( [ comment.get( 'ID' ), 'data' ], comment );
+			}
+
+			if ( comment.get( 'parent' ) ) {
+				commentsTree.updateIn( [ comment.getIn( [ 'parent', 'ID' ] ), 'children' ], ( children ) => children.push( comment.get( 'ID' ) ) );
+			}
+
+			// We check here for commentNodeChanged in order to not add the comment if we already saw it
+			if ( commentsTree.getIn( [ comment.get( 'ID' ), 'parentId' ] ) === null ) {
+				commentsTree.update( 'children', ( children ) => children.push( comment.get( 'ID' ) ) );
+			}
+		} );// forEach comments
+	} ); // withMutation
+}
+
+/***
+ * Finds and modifies items in a list
+ *
+ * @param {Immutable.List} list the list in which find and modify
+ * @param {Function} predicate function that returns true for an item to modify
+ * @param {Function} updater function that returns new value for item
+ * @param {Boolean} updateAll whether update all matches of predicate or just the first
+ * @returns {Immutable.List} modified list
+ */
+export function listFindAndModify( list, predicate, updater, updateAll = false ) {
```

Comment:
> You're suggesting remove the multi case?

If we're not using it, yeah. No reason to have it.

> It's still 2 not chain-able method calls I need to encapsulate in some function.

Having the function is fine. Just thinking we can simplify it and use more idiomatic immutable methods.

Maybe for a name `updateExistingIn`? Given how Immutable works, this could really work for List or Map.. It's pretty much like `updateIn`, but it wouldn't create missing keys.